tools/gen_cli_docs.py

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import pathlib
+import sys
+
+from craft_cli.dispatcher import Dispatcher, _CustomArgumentParser
+
+this_dir = pathlib.Path(os.path.split(__file__)[0])
+sys.path.insert(0, str((this_dir / "..").absolute()))
+
+from charmcraft import application, services
+
+
+def command_page_header(cmd, options_str, required_str):
+    underline = "=" * len(cmd.name)
+    overview = fix_spelling(cmd.overview)
+    return f"""
+.. _ref_commands_{cmd.name}:
+
+{cmd.name}
+{underline}
+{overview}
+
+Usage
+-----
+
+:command:`charmcraft {cmd.name}{options_str}{required_str}`
+
+"""
+
+
+# Temporary fix for spelling checks.
+def fix_spelling(t):
+    for old, new in [("artifact", "artefact"), ("Initialize", "Initialise")]:
+        t = t.replace(old, new)
+    return t
+
+
+def make_sentence(t):
+    return t[:1].upper() + t[1:].rstrip(".") + "."
+
+
+def not_none(*args):
+    return [x for x in args if x != None]
+
+
+def make_section(title, items):
+    s = ""
+    if items:
+        underline = "-" * len(title)
+        s = f"{title}\n{underline}\n\n"
+
+    for dest, (names, help_str) in sorted(items):
+        names = " or ".join([f"``{name}``" for name in names])
+        description = make_sentence(fix_spelling(help_str))
+        s += f"{names}\n   {description}\n"
+
+    s += "\n"
+    return s
+
+
+def main(docs_dir):
+    """Generate reference documentation for the command line interface,
+    creating pages in the docs/reference/commands directory and creating the
+    directory itself if necessary."""
+
+    # Create the directory for the commands reference.
+    commands_ref_dir = docs_dir / "reference" / "commands"
+    if not commands_ref_dir.exists():
+        commands_ref_dir.mkdir()
+
+    # Create a dispatcher like Charmcraft does to get access to the same options.
+    charmcraft_services = services.CharmcraftServiceFactory(app=application.APP_METADATA)
+    app = application.Charmcraft(app=application.APP_METADATA, services=charmcraft_services)
+    application.commands.fill_command_groups(app)
+    command_groups = app.command_groups
```

Comment:
This section is modified to generate Charmcraft's command groups.

@tigarmo should we move this file into craft-cli or craft-application?

Reply:
+1 on upstreaming with a slight preference for craft-application

Snapcraft has a similar but diverged version of this file too, so it's time to DRY (or make an issue for it).

Reply:
I think so! 

> craft-cli or craft-application

maybe both (some parts in -cli and some in -application)

Reply:
OK, I've added an item to our sync meeting to discuss long-term options